src/cascadia/TerminalApp/Pane.cpp

```diff
@@ -214,48 +214,172 @@ bool Pane::ResizePane(const ResizeDirection& direction)
 // Return Value:
-// - true if we or a child handled this focus move request.
-bool Pane::NavigateFocus(const FocusDirection& direction)
+// - The result of navigating from source according to direction, which may be
+//   nullptr (i.e. no pane was found in that direction).
+std::shared_ptr<Pane> Pane::NavigateDirection(const std::shared_ptr<Pane> sourcePane, const FocusDirection& direction)
 {
-    // If we're a leaf, do nothing. We can't possibly have a descendant with a
-    // separator the correct direction.
+    // Can't navigate anywhere if we are a leaf
     if (_IsLeaf())
     {
-        return false;
+        return nullptr;
+    }
+
+    // If the MRU previous pane is requested we can't move; the tab handles MRU
+    if (direction == FocusDirection::None || direction == FocusDirection::Previous)
+    {
+        return nullptr;
     }
 
-    // If the focus direction does not match the split direction, the focused pane
+    // Check if we in-order traversal is requested
+    if (direction == FocusDirection::NextInOrder)
+    {
+        return NextPane(sourcePane);
+    }
+
+    if (direction == FocusDirection::PreviousInOrder)
+    {
+        return PreviousPane(sourcePane);
+    }
+
+    // We are left with directional traversal now
+    // If the focus direction does not match the split direction, the source pane
     // and its neighbor must necessarily be contained within the same child.
     if (!DirectionMatchesSplit(direction, _splitState))
     {
-        return _firstChild->NavigateFocus(direction) || _secondChild->NavigateFocus(direction);
+        if (auto p = _firstChild->NavigateDirection(sourcePane, direction))
+        {
+            return p;
+        }
+        return _secondChild->NavigateDirection(sourcePane, direction);
     }
 
     // Since the direction is the same as our split, it is possible that we must
     // move focus from from one child to another child.
     // We now must keep track of state while we recurse.
-    auto focusNeighborPair = _FindFocusAndNeighbor(direction, { 0, 0 });
+    const auto paneNeighborPair = _FindPaneAndNeighbor(sourcePane, direction, { 0, 0 });
 
-    // Once we have found the focused pane and its neighbor, wherever they may
-    // be we can update the focus.
-    if (focusNeighborPair.focus && focusNeighborPair.neighbor)
+    if (paneNeighborPair.source && paneNeighborPair.neighbor)
     {
-        focusNeighborPair.neighbor->_FocusFirstChild();
-        return true;
+        return paneNeighborPair.neighbor;
     }
 
-    return false;
+    return nullptr;
+}
+
+// Method Description:
+// - Attempts to find the succeeding pane of the provided pane.
+// - NB: If targetPane is not a leaf, then this will return one of its children.
+// Arguments:
+// - targetPane: The pane to search for.
+// Return Value:
+// - The next pane in tree order after the target pane (if found)
+std::shared_ptr<Pane> Pane::NextPane(const std::shared_ptr<Pane> targetPane)

```

Comment:
though, I suppose tmux does this with `select-pane -t +/-/!`:

```
     target-pane (or src-pane or dst-pane) may be a pane ID or takes a
     similar form to target-window but with the optional addition of a
     period followed by a pane index or pane ID, for example:
     ‘mysession:mywindow.1’.  If the pane index is omitted, the
     currently active pane in the specified window is used.  The
     following special tokens are available for the pane index:

     Token                  Meaning
     {last}            !    The last (previously active) pane
     {next}            +    The next pane by number
     {previous}        -    The previous pane by number
     ...
```

THOUGH, for tmux, the pane location always stays the same, they just move the content. So doing a `swap-pane -s 2 -t 0` will only move the content between 0 and 2, not actually move the id's themselves.

That would mean that the ID order in tmux is always the same as the structural order.

Reply:
Well I'm glad you convinced yourself :). I chose structural order, as you noticed, specifically because the ID order isn't stable and I thought it would be confusing to have the focus move "randomly."

Reply:
I just look forward to when you get a feature request where someone wants custom navigation and you have to implement user-specified continuations to support it ;)